### ISSUE TITLE
Feature dimension not checking out

### ISSUE THREAD

**Issue by p-CXTN** [LeonResearch/City-Networks#2]
Hi, thanks for the work and the dataset.

However, I'm confused with the dimension of the (augmented) node features. 

In the description in your paper in Appendix B.1, you take top 8 classes of the categorical features (node & edge,  land use, lanes, and road type), which by my understanding should result in 9 dimensioned one-hot vectors for each? Then adding the 3 numerical features for nodes, 2 numerical features for edges, and 2 binary features (one-hot encoded, IMO should be 2 dimensions each), $3\times 9+3+2+2\times 2=36$, which is 1 dimension short of your stated 37 dimensions.

When I check the actual features in the dataset, it seems like the last 10 dimensions should be an averaged one-hot encoded categorical feature (and according to your code, should correspond to class "highway", and I guess this is the class "road type" described in your paper?). But I didn't find any evidence in your code/paper that confirms this 10-dimensional one-hot encoding. Is this the case?

I'm also a bit confused with the 2 binary features, which should be encoded to 2-dimensional one-hot vectors. But in your code [here](https://github.com/LeonResearch/City-Networks/blob/main/dataset_generation/generate_network.py#L133C1-L139C2) (lines from 133 onwards)

```python
top_k_dict = {
    "oneway": 4,
    "lanes": 8, 
    "reversed": 4,
    "highway": 8,
    "landuse":8,
}
```

the top k classes of both binary features are chosen to be 4. Could you confirm that this doesn't matter and create a 2-dimensional feature in the end?

Thanks in advance for your clarification.

**Comment by LeonResearch** [comment]
Dear p-CXTN,

Thank you for your interest in our work! 

Your explanation is right on point about this issue! I just reran the dataset generation code on Paris and got the following result (see screenshots below). 

- For the "reversed" feature, except True and False, there is another value [True, False], which means there are multiple roads of different "reversed" types between these two nodes (OSMnx uses nx.MultiDiGraph for city maps). As top_k_dict["reversed"] = 4, we will have 3 dimensions for this feature after one-hot encoding. This explains the 1 extra dimension in your question, and we will make this information clear in our revised paper.

- For "oneway", there were only two values, True and False, when we constructed our dataset in early 2025. However, when I reran the generation code, it returned another value [True, False] with only 1 entry for Paris. Note that this is because OSMnx queries the latest OpenStreetMap, which means that even if we fix its version, it will still return us different city maps if there is an information update on OSM before our second query.

<img width="722" height="282" alt="Image" src="https://github.com/user-attachments/assets/1e1993b9-0252-4d68-b761-f29722f7a1e1" />

<img width="733" height="162" alt="Image" src="https://github.com/user-attachments/assets/05b00f72-3be6-44c4-8d92-149dd6cafc73" />

- For the augmented edge features, the code is at line 99 in `generate_dataset.py`:
```python
if args.augment_node_attr:
    # Create empty tensor for sum and count
    node_edge_attr_sum = torch.zeros(node_features.shape[0], edge_features.shape[1])                          
    node_edge_count = torch.zeros(node_features.shape[0], 1)
    # sum the corresponding edge attr for each node
    node_edge_attr_sum.index_add_(0, edge_indices[1], edge_features)
    node_edge_count.index_add_(
        0,
        edge_indices[1],
        torch.ones_like(edge_indices[1], dtype=torch.float).view(-1, 1)
    )
    # Prevent division by zero for nodes with no neighbors if the graph is not connected
    node_edge_count[node_edge_count == 0] = 1.0
    # Compute the mean of the neighbouring edge attributes for each node
    node_edge_attr_mean = node_edge_attr_sum / node_edge_count
    # Concatenate the original node features with the averaged edge attributes
    x_augmented = torch.cat([node_features, node_edge_attr_mean], dim=1)
    torch.save(x_augmented, f'{data_dir}node_features_augmented.pt')
```
And they are stacked in this order:
```python
edge_features = torch.tensor(
    np.hstack(
        (edge_df[edge_numerical_features].values, edge_features_categorical)
    ),
    dtype=torch.float32
)
```
where `edge_categorical_features = ["oneway", "lanes", "reversed", "highway"]`. It means that the last 9 dimensions correspond to "highway", if you sum these values up `paris[0].x[:,-9:].sum(dim=1)` the result will be `1.0` for all nodes. 



I hope this helps answer your question :)

Best,
Huidong

**Comment by p-CXTN** [comment]
Thank you for your clarification!
